core/src/main/scala/org/ensime/indexer/domain.scala

```diff
@@ -125,6 +128,115 @@ final case class MethodName(
   def fqnString = owner.fqnString + "." + name + descriptor.descriptorString
 }
 
+/**
+ * Generics signature
+ */
+
+trait GenericArg extends FullyQualifiedName
+trait SignatureType extends FullyQualifiedName
+
+case class GenericClass(
+    genericParam: Seq[GenericParam],
+    superClasses: Seq[GenericClassName]
+) extends SignatureType {
+
+  override def contains(o: FullyQualifiedName): Boolean = o match {
+    case g: GenericClass => g == this
+    case sthElse =>
+      genericParam.exists(_.contains(sthElse)) ||
+        superClasses.exists(_.contains(sthElse))
+  }
+
+  override def fqnString: String = {
+    "<" + genericParam.map(_.fqnString).mkString + ">" +
+      superClasses.map(_.fqnString).mkString
+  }
+}
+
+case class GenericParam(
+    name: String,
+    classNames: Seq[GenericClassName]
+) extends SignatureType {
+
+  override def contains(o: FullyQualifiedName): Boolean = o match {
+    case gp: GenericParam => this == gp
+    case sthElse => classNames.exists(_.contains(o))
+  }
+
+  override def fqnString: String = {
+    classNames.map(_.fqnString).mkString
+  }
+}
+
+case class GenericClassName(
+    className: ClassName,
```

Comment:
instead of redefining `ClassName`, it would be better to add extra fields here. Everything in `FullyQualifiedName` is representative of the bytecode representation and generics are erased by that time, so all generic information needs to be captured here in the `GenericClass` hierarchy.